### PR TITLE
add simple colormap object

### LINKED ISSUE #862: Generic colormap representation
## 🚀 Feature
Right now in our model files we use the vispy colormap object. It would be nice to completely separate our model files from vispy as noted here https://github.com/napari/napari/pull/686#discussion_r363550535 in discussions around serializing the viewer and round-trips.

After #782 (which does a similar refactor but for colors) and #854 (which adds a camera model) both get merged our models will be vispy independent.

Note I'm not sure yet what this representation should be or if vispy colormaps are serializable as they are now, but am making this issue to discuss this.

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
This PR adds a simple numpy based colormap object that can easily be converted to a vispy or other colormap. It superceeds #1517. Will close #862 

It is still a WIP, I will update as I go. So far I've only looked at image/ surface properly. Labels, points, vectors, shapes still need work.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by sofroniewn** [comment]
Just did some more work on this, it's getting close. I now have to colormap conversion working for all layers and a test that ensure parity between our colormaps and a vispy colormap object to 6 decimals on all our available colormaps.

I need to fix some tests which are now basic syntax errors and update some doc strings. I might also work on some additional ways to accept colormaps, and I need to fix the serialization tests.

**Comment by sofroniewn** [comment]
@jni @kevinyamauchi @tlambert03 this PR is now ready for review.

It adds a simple Colormap object capable of mapping colors with simple numpy operations. I've preserved the API so that all vispy Colormap objects can still be passed. We might have problems with some vispy colormaps that are purely shader defined (like `grays` and `fire`, but they are not common I think).

I'm also still just putting `colormap.name` inside the layer data tuple so the serialization tests pass, but fixing that would be easy to address in a new PR.

I noticed that @quantumjot was doing things with custom colormaps here https://github.com/quantumjot/arboretum/blob/master/arboretum/_colormaps.py for the Tracks Layer work and maybe in #1361. I'm curious if this work would still allow your use cases?

**Comment by quantumjot** [comment]
Hi @sofroniewn - this looks great! From a cursory look, I don't see why this will cause any problems for the track layer going forward, and I think having a unified colormap object makes things easier in general. Only thing I didn't see is a color cycle method for discrete colormaps, but maybe I missed that.

**Comment by sofroniewn** [comment]
> Hi @sofroniewn - this looks great! From a cursory look, I don't see why this will cause any problems for the track layer going forward, and I think having a unified colormap object makes things easier in general

That's great, thanks for the quick look @quantumjot!

> Only thing I didn't see is a color cycle method for discrete colormaps, but maybe I missed that.

There is a way to make a discrete colormap with n colors in n+1 bins using the `interpolation='zero'` option. Right now the API there matches the vispy one identically, but we could add more convenience features here if desired in the future

**Comment by kevinyamauchi** [comment]
Sorry, things have been crazy at work! I can review on Sunday.

File: napari/layers/shapes/_tests/test_shapes.py
Comment (kevinyamauchi) [inline napari/layers/shapes/_tests/test_shapes.py:1125]: I think it's neat that this works now!

File: napari/utils/colormaps/colormaps.py
Comment (kevinyamauchi) [inline napari/utils/colormaps/colormaps.py]: Since this function has a lot of branches, I think it would be nice to add some tests.

File: napari/layers/labels/_tests/test_labels.py
Comment (kevinyamauchi) [inline napari/layers/labels/_tests/test_labels.py]: I'm not sure if we have plans to subclass `Colormap`, but should we use `isinstance` instead?

```suggestion
    assert isinstance(layer.colormap, Colormap)
```

File: napari/_vispy/vispy_surface_layer.py
Comment (kevinyamauchi) [inline napari/_vispy/vispy_surface_layer.py]: I think it's a bit confusing that we now have 2 types of `Colormap` objects now (vispy and napari). Perhaps we can import the vispy colormap as `VispyColormap`?

File: napari/utils/colormaps/colormaps.py
Comment (kevinyamauchi) [inline napari/utils/colormaps/colormaps.py]:
```suggestion
def ensure_colormap(colormap: ValidColormapArg) -> Colormap:
```

It looks like just the `Colormap` object is returned now. If so, I think the `Returns` section of the docstring needs to be updated as well.

File: napari/utils/colormaps/colormap.py
Comment (kevinyamauchi) [inline napari/utils/colormaps/colormap.py]: This is a bit of a nitpick, but I think `undefined` is a bit of a confusing name since the color map is defined. Perhaps something like `custom_colormap` instead?

File: napari/utils/colormaps/colormaps.py
Comment (kevinyamauchi) [inline napari/utils/colormaps/colormaps.py]: I think the docstring needs to be updated now that the returned colormap is a napari `Colormap`.

File: napari/layers/image/image.py
Comment (kevinyamauchi) [inline napari/layers/image/image.py:465]: What do you think about storing a dictionary with the kwargs required to make the `Colormap` object? That way even custom colormaps could be reloaded from the `layer._get_state()` output.

File: napari/utils/colormaps/colormap.py
Comment (kevinyamauchi) [inline napari/utils/colormaps/colormap.py]: I believe we include a definition of the enum values in our other mode enums. I think it would be nice to do so here as well.

**Review by kevinyamauchi — COMMENTED** [review]
I think this is looking good, @sofroniewn ! I have left some comments below.

Also, I was wondering if we should rename `/napari/utils/colormaps/colormaps.py`. It's a bit confusing because there's also a `colormaps.py` in the same folder. Maybe we can name it `colormap_utils.py` instead?

File: napari/_vispy/vispy_surface_layer.py
Comment (tlambert03) [reply napari/_vispy/vispy_surface_layer.py]: was just about to make the same comment on the other vispy file :)

File: napari/_vispy/vispy_surface_layer.py
Comment (tlambert03) [inline napari/_vispy/vispy_surface_layer.py:56]: I think this check is no longer necessary after #1456, now that gamma is performed by the shader?  Can you check to see if it works to just assign `cmap = Colormap(*self.layer.colormap)` without the gamma check?

File: napari/layers/image/image.py
Comment (tlambert03) [reply napari/layers/image/image.py:465]: as briefly discussed before... I was kinda hoping this would actually just be the colormap object.  While that may not be immediately serializable (which I know was the original intended use for `_get_state`), truth is we've never actually used it that way, and now it forms a main part of the layer_data_tuple that plugins might get.  If we give them the object, they immediately have something useful.  If we just give them the name (or stuff to construct it), then they have to either depend on some of our convenience functions, or recreate them.  Serialization can be solved if/when we've ever truly need it, but passing the actual object would be immediately useful in (e.g.) napari-svg right?

File: napari/layers/labels/labels.py
Comment (tlambert03) [inline napari/layers/labels/labels.py:580]: ❤️

File: napari/layers/surface/surface.py
Comment (tlambert03) [inline napari/layers/surface/surface.py:240]: same here... would love to see this just be `self.colormap`

File: napari/utils/colormaps/colormaps.py
Comment (tlambert03) [inline napari/utils/colormaps/colormaps.py]: just personal preference... but since this is basically just a modified `Colormap` constructor, I think it would be lovely for it to live on the `Colormap` (or at least be a private function if you really want it outside the class):

```python
class Colormap:
    @classmethod
    def from_vispy(cls, colormap, name=None):
        ...
        return cls(
            name=name,
            colors=colormap.colors.rgba,
            controls=controls,
            interpolation=interpolation,
        )
```

that cleans up imports elsewhere, so it's always:
```python
from napari.utils import Colormap

cmap = Colormap.from_vispy(vispy_colormap)
```
rather than sometimes importing Colormap and other times importing `convert_vispy_colormap` depending on the starting material.

also, having `name=None` here is going to break the typing on our Colormap object.  for example, if a name is not provided, and someone tries to do something like `cmap.name.lower()`, it would now be an attribute error. Either add `name=vispy_colormap` or something here... or make the default of `Colormap` be `None` and then set the default to `custom_colormap` in the `Colormap.__init__` function

File: napari/utils/colormaps/colormaps.py
Comment (tlambert03) [inline napari/utils/colormaps/colormaps.py:82]: I think it would be reasonable to just check `isinstance(colormap, VispyColormap)` and throw a `TypeError` otherwise, rather than checking hasattr.  This function is, afterall, dedicated to vispy objects.

File: napari/utils/colormaps/colormap.py
Comment (tlambert03) [inline napari/utils/colormaps/colormap.py]: lowercase?

**Review by tlambert03 — APPROVED** [review]
really great work @sofroniewn!  love what this does to a lot of our code.
I'm marking as "approve" but I'd definitely like to hear your thoughts on some of my comments before merge.

**Comment by sofroniewn** [comment]
Thanks for review @kevinyamauchi @tlambert03, I'll look into addressing concerns promptly

File: napari/utils/colormaps/colormaps.py
Comment (kevinyamauchi) [reply napari/utils/colormaps/colormaps.py]: I think making a `Colormap.from_vispy()` constructor is a good idea!

File: napari/_vispy/vispy_surface_layer.py
Comment (sofroniewn) [reply napari/_vispy/vispy_surface_layer.py:56]: The surface layer uses the `Mesh` visual from vispy, which we didn't vendor in #1456. Not sure if gamma has been added to Mesh upstream in vispy, but I just tried removing the gamma check and it didn't work.

File: napari/_vispy/vispy_surface_layer.py
Comment (sofroniewn) [reply napari/_vispy/vispy_surface_layer.py]: done in both files

File: napari/layers/labels/_tests/test_labels.py
Comment (sofroniewn) [reply napari/layers/labels/_tests/test_labels.py]: fixed here and everywhere

File: napari/utils/colormaps/colormap.py
Comment (sofroniewn) [reply napari/utils/colormaps/colormap.py]: i've gone with `custom`

File: napari/utils/colormaps/colormap.py
Comment (sofroniewn) [reply napari/utils/colormaps/colormap.py]: gone with `n_controls`

File: napari/utils/colormaps/colormaps.py
Comment (sofroniewn) [reply napari/utils/colormaps/colormaps.py:82]: I've added in the type check, unfortunately I think I need to add the `hasattr` check as I have encountered vispy colormaps that don't have those attrs! They may have been strangely constructed ones, like using mostly custom shader code, which we might not be able to translate properly anyway, but the hasattr check seemed safest ....

File: napari/utils/colormaps/colormaps.py
Comment (sofroniewn) [reply napari/utils/colormaps/colormaps.py]: I've changed the default `name='vispy'` in `convert_vispy_colormap`, but I think I'd rather leave the `from_vispy` method off of the `Colormap` object. We only use `convert_vispy_colormap` in one file in the code base, the one it is currently defined in, and I wouldn't really want to encourage users to me making vispy colormaps and then converting them to napari ones (note we can accept vispy colormaps fine through the API).

File: napari/utils/colormaps/colormaps.py
Comment (sofroniewn) [reply napari/utils/colormaps/colormaps.py]: done

File: napari/utils/colormaps/colormaps.py
Comment (sofroniewn) [reply napari/utils/colormaps/colormaps.py]: Done

File: napari/layers/image/image.py
Comment (sofroniewn) [reply napari/layers/image/image.py:465]: I'd like to wait on this until after this PR and make a dedicated issue to follow up. Making that change will require additional PRs to napari-svg to be able to use the new return value and it could get tricky to coordinate all the PRs across repos. As the current change is doesn't actually change any functionality, I'd like to get this in first and then fix, but agreed it would be nice to resolve

File: napari/layers/surface/surface.py
Comment (sofroniewn) [reply napari/layers/surface/surface.py:240]: Agreed, see comment above about staging this into multiple PRs so as not to break napari-svg

File: napari/utils/colormaps/colormaps.py
Comment (sofroniewn) [reply napari/utils/colormaps/colormaps.py]: Comprehensive tests added, and existing tests consolidate into `napari.utils.colormaps` folder

**Review by sofroniewn — COMMENTED** [review]
Ok, I have gone through and addressed all the reviewer comments apart from two.

I have not added the `from_vispy` method to the Colormap object we only need it in one file, see comments above.

I have not made the `_get_state` dict get the whole colormap, just the `colormap.name` as giving the colormap now would be a breaking change and require a series of PRs with napari-svg to ensure everything works. In the interest of keeping our PRs small and manageable and the project moving along I'd like us to merge as is now which is fully backwards compatible, and then make breaking changes in a new PR. I will make a dedicated issue for that if this gets merged as is so we don't forget.

@jni if you could give this a quick pass that would be great, it has already received two detailed reviews and an approval, but maybe you can focus on changes since the last reviews in case anything funny crept in. Thanks!!

File: napari/_qt/layer_controls/qt_labels_controls.py
Comment (jni) [inline napari/_qt/layer_controls/qt_labels_controls.py:498]: Can you explain this change? Why mix in opacity now?

File: napari/components/_tests/test_multichannel.py
Comment (jni) [inline napari/components/_tests/test_multichannel.py:20]: Sorry, this is nitpicky, but why isn't this simple_colormaps like the first two?

File: napari/layers/intensity_mixin.py
Comment (jni) [inline napari/layers/intensity_mixin.py]: napari

File: napari/utils/colormaps/_tests/test_colormap.py
Comment (jni) [inline napari/utils/colormaps/_tests/test_colormap.py]: cmap?

File: napari/utils/colormaps/_tests/test_colormaps.py
Comment (jni) [inline napari/utils/colormaps/_tests/test_colormaps.py]: Should there be a ]?

File: napari/utils/colormaps/colormap_utils.py
Comment (jni) [inline napari/utils/colormaps/colormap_utils.py]: Interpolation

File: napari/utils/colormaps/colormap_utils.py
Comment (jni) [inline napari/utils/colormaps/colormap_utils.py:384]: @sofroniewn it looks to me like you need to update the definition of ValidColormapArg but haven't done so?

**Review by jni — APPROVED** [review]
@sofroniewn most of my comments are minor. Things I'm worried about:

- the type logic in ensure_colormap is pretty messy right now. Which is not a good prelude to:
- on reading this, my overarching feeling is that it's really sad to be adding yet another Colormap object to the ecosystem. I just saw that matplotlib colormaps do have a name attribute. But, we don't want to depend on matplotib. Fine. But it would be great if we could *accept* matplotlib colormaps as input. It would also be great if our colormaps could mimic the matplotlib colormap API, so that our colormaps could (conceivably) be used by matplotlib.
- of course, all this can come later and shouldn't block this PR. Would you mind creating an issue about this so we can follow up?

File: napari/_qt/layer_controls/qt_labels_controls.py
Comment (sofroniewn) [reply napari/_qt/layer_controls/qt_labels_controls.py:498]: This change is definitely independent of this PR, and could be done separately, but it makes the color of the swatch match the color of the label in the canvas and the color in the thumbnail, which I think will be nicer for people working with labels at lower opacity values

File: napari/components/_tests/test_multichannel.py
Comment (sofroniewn) [reply napari/components/_tests/test_multichannel.py:20]: I have made `SIMPLE_COLORMAPS` for consistency, I think reasons were purely historical, good catch

File: napari/layers/intensity_mixin.py
Comment (sofroniewn) [reply napari/layers/intensity_mixin.py]: done

File: napari/utils/colormaps/_tests/test_colormap.py
Comment (sofroniewn) [reply napari/utils/colormaps/_tests/test_colormap.py]: done

File: napari/utils/colormaps/_tests/test_colormaps.py
Comment (sofroniewn) [reply napari/utils/colormaps/_tests/test_colormaps.py]: done

File: napari/utils/colormaps/colormap_utils.py
Comment (sofroniewn) [reply napari/utils/colormaps/colormap_utils.py]: done

File: napari/utils/colormaps/colormap_utils.py
Comment (sofroniewn) [reply napari/utils/colormaps/colormap_utils.py:384]: I have added, but had to move type definition into colormaps folder to avoid circular imports. This is probably not a long term ideal solution but seems like a reasonable step for now

**Comment by sofroniewn** [comment]
> - the type logic in ensure_colormap is pretty messy right now. Which is not a good prelude to:
> - on reading this, my overarching feeling is that it's really sad to be adding yet another Colormap object to the ecosystem. I just saw that matplotlib colormaps do have a name attribute. But, we don't want to depend on matplotib. Fine. But it would be great if we could accept matplotlib colormaps as input. It would also be great if our colormaps could mimic the matplotlib colormap API, so that our colormaps could (conceivably) be used by matplotlib.
> - of course, all this can come later and shouldn't block this PR. Would you mind creating an issue about this so we can follow up?

I completely agree the `ensure_colormap` is quite messy right now, and that is mainly for a desire to be backwards compatible right now. We could deprecate some of these more esoteric inputs like a dictionary of named vispy colormaps, which I don't think is used anywhere in the codebase, and fairly redundant with named tuples. We could then keep only the simplest options and add support for all matplotlib colormaps, which has been requested in #899, and make it clear how to convert our colormaps to matplotlib colormaps (which is a one-linear for each of the `linear` and `zero` options)

```python
# for linear interpolated napari colormap
from matplotlib.colors import LinearSegmentedColormap
mp_cmap = LinearSegmentedColormap.from_list(cmap.name, cmap.map(np.linspace(0, 1, 256)))
```

```python
# for zero interpolated napari colormap
from matplotlib.colors import ListedColormap
mp_cmap = ListedColormap(cmap.map(np.linspace(0, 1, 256)), name=cmap.name)
```

See here using converting the napari `turbo` colormap to matplotlib and using inside the console:

Screen Shot 2020-08-23 at 1 17 15 PM

I will create a new issue referencing this and #899 to continue discussion, I really think this barebones colormap object will help us unify matplotlib and vispy colormaps under-the-hood and support both.

**Comment by sofroniewn** [comment]
Issue created here https://github.com/napari/napari/issues/1575. All tests are passing, we can make fixes in followup PRs and continue discussion there!!

**Pull request merged**
